### ISSUE TITLE
'Continue L;' statement does not cancel stream subscription

### ISSUE THREAD

**Issue by alsemenov** [dart-lang/sdk#25310]
The Dart Programming Language Speciﬁcation version 1.11, chapter 17.15 Continue reads:
>Let sc be a continue statement. If sc is of the form continue L;, then let sE
>be the the innermost labeled do (17.8), for (17.6) or while (17.7) statement or
>case clause with label L enclosing sc . If s c is of the form continue; then let sE
>be the the innermost do (17.8), for (17.6) or while (17.7) statement enclosing sc .
> ...
> If sE is an asynchronous for loop (17.6.3), let ak be the set of asynchronous
>for loops and yield-each statements (17.16.2) enclosing sc that are enclosed
>in sE , 1 ≤ k ≤ m, where a k is enclosed in ak+1 . The stream subscriptions
>associated with aj are canceled, 1 ≤ j ≤ m, innermost ﬁrst, so that aj is
>canceled before aj+1 .

The following code demonstrates that stream subscription is not canceled:
```
import 'dart:async';

class WrapStream extends Stream {
  int _id;
  Stream _stream;
  WrapStream(this._id, this._stream);
  listen(onData, {onError,onDone,cancelOnError}) {
    print("listen $_id");
    return new _WrapSub(_stream.listen(onData, onError:onError, onDone:onDone,
        cancelOnError: cancelOnError), _id);
  }
}

class _WrapSub implements StreamSubscription {
  int _id;
  StreamSubscription _sub;
  _WrapSub(this._sub, this._id);
  void onData(d) { _sub.onData(d); }
  void onError(d) { _sub.onError(d); }
  void onDone(d) { _sub.onDone(d); }
  void pause([Future f]) { _sub.pause(f); }
  void resume() { _sub.resume(); }
  bool get isPaused => _sub.isPaused;
  Future cancel() {
    print("cancel $_id");
    var f = _sub.cancel();
    if (f is Future) {
      f = f.whenComplete(() { print("cancelled! $_id"); });
    }
    return f;
  }
  Future asFuture([val]) => _sub.asFuture(val);
}

Stream makeStream(Iterable iterable, int id) {

Stream s = new Stream.fromIterable(iterable);
print("created stream $id");
return new WrapStream(id, s);
}

Future test5() async {
  List<int> log = [];
  int i = 0;
  L:
  await for (String s in makeStream(['a0', 'b0'], i++)) {
    switch (s) {
      case 'a0':
        print('case a0');
        await for (String s in makeStream(['a1', 'b1', 'c1'], i++)) {
          print('s: $s');
          switch (s) {
            case 'a1':
              print('case a1');
              continue;
            case 'b1':
              print('case b1');
              continue M;
            default:
              print("case default $s");
              break;
          }
        }
        break;

      M:
      case 'b0':
        print("label M");
        break;
    }
  }
}

main() {
  test5().then((_) => print("test1 completed"));
}
```

Test produces output:
>created stream 0
>listen 0
>case a0
>created stream 1
>listen 1
>s: a1
>case a1
>s: b1
>case b1
>label M
>label M
>test1 completed

Expected output is:
>created stream 0
>listen 0
>case a0
>created stream 1
>listen 1
>s: a1
>case a1
>s: b1
>case b1
>**cancel 1**
>label M
>label M
>test1 completed

Dart VM version: 1.13.1 (Thu Dec 17 06:40:17 2015) on "windows_x64"

**Comment by lrhn** [comment]
Likely caused by the one-element cache on the `StreamIterator` used internally in the implementation of the await-for. That means that the inner stream has seen the next (and last) element and is therefore already done at the time the cancel happens, and if the `StreamIterator` knows this, it won't forward the `cancel` method.

If you add a fourth element to the inner stream, I expect you will see a cancel.

That's an explanation, not an excuse, the behavior is not spec compliant. (Personally I think the specification is overly specific at this point, disallowing the current behavior, but it is unlikely to change, so we might need to use a different helper class than the `StreamIterator`).

**Comment by alsemenov** [comment]
I checked on Dart VM version: 1.13.2 (Tue Jan 05 06:46:24 2016) on "windows_x64" the problem is reproducible, even if I add more elements to inner stream.

**Comment by mhausner** [comment]
CL https://codereview.chromium.org/2030763002/ fixes this. The close/cancel calls in the implicit finally clauses were not properly inlined for forward jumps to outer switch statements.

**Comment by mhausner** [comment]
Fixed in 1859ce46c337e4e047e6543c561ff77af7910074
